xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -347,26 +353,32 @@ static boolean matchHostName(String hostName, String pattern) {
   private final class ConfigSelector extends InternalConfigSelector {
     @Override
     public Result selectConfig(PickSubchannelArgs args) {
-      // Index ASCII headers by keys.
-      Map<String, Iterable<String>> asciiHeaders = new HashMap<>();
+      RoutingConfig routingCfg = routingConfig;
```

Comment:
As I mentioned in the HttpFilter PR, that approach is wrong. The data that `selectConfig()` uses can be stale but it has to be consistent. Otherwise, it may cause weird behavior that is extremely hard to debug.

Reply:
Fixed. 

This is more like optimistic locking. The do-while loop is intended to avoid the race between a cluster being selected and its reference being deleted due to xDS updates. If the selected cluster cannot be retained, the list of Routes must have been updated and we should try to redo the selection. Reading the latest list of Routes should be the beginning of cluster selection.